src/twisted/mail/test/test_smtp.py

```diff
@@ -669,6 +671,47 @@ def check(ignored):
 
         return self.loopback(server, client).addCallback(check)
 
+    def test_ESMTPSenderFactory_TLSError(self):
+        """
+        Attempting to connect to an ESMTP server which presents an invalid certificate
+        should produce a meaningful error.
+        """
+        # set up a dummy ESMTP server which will present a self-signed cert after
+        # STARTTLS
+        buildServerProtocol = lambda: DummyESMTP(contextFactory=ServerTLSContext())
+        serverFactory = Factory.forProtocol(buildServerProtocol)
+        serverPort = reactor.listenTCP(0, serverFactory, interface="127.0.0.1")
+        self.addCleanup(serverPort.stopListening)
+
+        # build a client, which won't accept the certificate presented by the dummy
+        # server.
+        sentDeferred = defer.Deferred()
+        clientFactory = smtp.ESMTPSenderFactory(
+            "username",
+            "password",
+            "source@address",
+            "recipient@address",
+            BytesIO(b"message"),
+            sentDeferred,
+            retries=0,
+            requireAuthentication=False,
+            contextFactory=optionsForClientTLS("testdomain"),
+        )
+
+        # connect the two together
+        connector = reactor.connectTCP(
+            serverPort.getHost().host, serverPort.getHost().port, clientFactory
+        )
+        self.addCleanup(connector.disconnect)
+
+        def checkVerifyFailedMessage(e: smtp.SMTPConnectError) -> None:
+            # check that the SMTPConnectError has a message relating to the TLS error
+            self.assertIn("certificate verify failed", str(e))
+
+        return self.assertFailure(sentDeferred, smtp.SMTPConnectError).addCallback(
+            checkVerifyFailedMessage
+        )
```

Comment:
FYI, this sort of real-networking test that returns a Deferred is really not the right way to test protocol-parsing logic like this; it does unnecessarily non-deterministic I/O with the operating system.  If you can, it should be rewritten to be more robust using `twisted.test.iosim.IOPump` or similar.  At the very least however, please remove the direct call to `listenTCP` and instead use the higher-level `self.loopback(...)` to at least abstract out the connection-making logic.

Reply:
Yeah. As I wrote [before](https://github.com/twisted/twisted/pull/1645#discussion_r682124969), I struggled to get this to work with `iosim` because `startTLS` wants to wrap a real file descriptor.

I had a look at making it work with `self.loopback`. It could be my failure to understand how to use `loopback`, but as far as I can tell, it wants to build its own `ClientFactory` (a `LoopbackClientFactory`), which seems like it makes it impossible to instantiate an `SMTPSenderFactory` (which is where the interesting changes are here).

Reply:
@exarkun [says](https://matrix.to/#/!TlViTjmLDIWAUxzxsO:libera.chat/$TsVkMpUZ44hJl8S_qDYJoILEBQvlCeZXqtGbwLd6xow?via=libera.chat&via=matrix.org&via=gitter.im)

> You mentioned that there are problems using iosim because startTLS expects a "real life file descriptor"?  iosim has built-in support for startTLS though, so I wonder what happened there.
>
> src/twisted/test/test_amp.py tests AMP's startTLS usage and uses iosim
> any of the tests that use `connectedServerAndClient`